GitPullRequest.Services/AzureDevOpsRepository.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using LibGit2Sharp;
+using LibGit2Sharp.Handlers;
+
+namespace GitPullRequest.Services
+{
+    internal class AzureDevOpsRepository : RemoteRepository
+    {
+        public AzureDevOpsRepository(IRepository repo, string remoteName)
+            : base(repo, remoteName)
+        {
+        }
+
+        public override int FindPullRequestForCanonicalName(string canonicalName)
+        {
+            var match = Regex.Match(canonicalName, "^refs/pull/([0-9]+)/merge$");
+            if (match.Success && int.TryParse(match.Groups[1].Value, out int number))
+            {
+                return number;
+            }
+
+            return -1;
+        }
+
+        protected override IDictionary<string, string> GetReferences(IRepository repo, string remoteName)
+        {
+            // for Azure DevOps we need to fetch PR branches so we can explore their history and get the commit before the automatic merge commit that is done on the server
+            repo.Network.Fetch(remoteName, new string[] { "+refs/pull/*/merge:refs/remotes/origin/pull/*/merge" }, new FetchOptions { CredentialsProvider = CreateCredentialsHandler() });
```

Comment:
I think we might be able to fetch without creating the remote branches by doing this:

```suggestion
            repo.Network.Fetch(remoteName, new string[] { "refs/pull/*/merge" }, new FetchOptions { CredentialsProvider = CreateCredentialsHandler() });
```

We need to do something similar in GitHub for Visual Studio when finding the merge base of a PR (in order to generate the correct PR diff):
https://github.com/github/VisualStudio/blob/a22bf5ea2867edac9ca1d757775173c107269ecb/src/GitHub.App/Services/GitClient.cs#L451-L463

Reply:
That doesn't work unfortunately because of the wildcard. My understanding is that fetching without a destination in the refspec just fetches to FETCH_HEAD but the wildcard means it will fetch multiple things, and they can't all go in FETCH_HEAD.

Reply:
That code you've linked to is essentially the same was what the first version of my code was doing - if the commit isn't find, fetch the remote branch and try to find it again.

Reply:
> ...the wildcard means it will fetch multiple things, and they can't all go in FETCH_HEAD, or perhaps more precisely FETCH_HEAD can't point to all of them.

Damn, that's a shame. Maybe we could fetch and then immediately delete the remote branches? You'd hope that deleting a branch would be a light weight operation.

Maybe I'm over worrying about leaving PR branches behind? 🤔

Reply:
I think there are way too many variables here to worry about:

* A user could have their git config set up to pull PR branches already, so we could be deleting something they want
* A user could have their git config set up to pull PR branches already, but to a different naming convention that used here
* A user (like me) could have their git config set up to prune on fetch, so every fetch will clean up the extra branches anyway

I guess we _could_ read the git config to check all of these things, since we don't care what the naming convention of the local branch is, and if they're already fetching PRs we don't need to, and if they're not we can prune afterwards? Though personally I quite like that its not pruning because it means the next run is going to be quicker.